SDWebImage/UIView+WebCache.m

```diff
@@ -52,31 +52,57 @@ - (void)sd_internalSetImageWithURL:(nullable NSURL *)url
             }
-            dispatch_main_async_safe(^{
-                if (!sself) {
-                    return;
-                }
-                if (image && (options & SDWebImageAvoidAutoSetImage) && completedBlock) {
-                    completedBlock(image, error, cacheType, url);
-                    return;
-                } else if (image) {
-                    [sself sd_setImage:image imageData:data basedOnClassOrViaCustomSetImageBlock:setImageBlock];
-                    [sself sd_setNeedsLayout];
-                } else {
-                    if ((options & SDWebImageDelayPlaceholder)) {
-                        [sself sd_setImage:placeholder imageData:nil basedOnClassOrViaCustomSetImageBlock:setImageBlock];
+            if ((image && (options & SDWebImageAvoidAutoSetImage)) || (!image && !(options & SDWebImageDelayPlaceholder))) {
+                dispatch_main_async_safe(^{
+                    if (completedBlock) {
+                        completedBlock(image, error, cacheType, url);
+                    }
+                });
+                // no need to call setImageBlock and return
+                return;
+            }
+            
+            UIImage *targetImage;
+            NSData *targetData;
+            if (image) {
+                targetImage = image;
+                targetData = data;
+            } else if (options & SDWebImageDelayPlaceholder) {
+                targetImage = placeholder;
+                targetData = nil;
+            }
+            if ([sself isKindOfClass:NSClassFromString(@"FLAnimatedImageView")]) {
+                // performance enhancement for `FLAnimatedImage`
+                dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
+                    [sself sd_setImage:targetImage imageData:targetData basedOnClassOrViaCustomSetImageBlock:setImageBlock];
+                    dispatch_main_async_safe(^{
+                        if (!sself) {
+                            return;
+                        }
                         [sself sd_setNeedsLayout];
+                        if (completedBlock && finished) {
+                            completedBlock(image, error, cacheType, url);
+                        }
+                    });
+                });
+            } else {
+                dispatch_main_async_safe(^{
+                    if (!sself) {
+                        return;
+                    }
+                    [sself sd_setImage:targetImage imageData:targetData basedOnClassOrViaCustomSetImageBlock:setImageBlock];
+                    [sself sd_setNeedsLayout];
+                    if (completedBlock && finished) {
+                        completedBlock(image, error, cacheType, url);
                     }
-                }
-                if (completedBlock && finished) {
-                    completedBlock(image, error, cacheType, url);
-                }
-            });
+                });
+            }
```

Comment:
to avoid repeating the code on if and else we can do something like:
```
dispatch_queue_t queueForSetImage = ([sself isKindOfClass:NSClassFromString(@"FLAnimatedImageView")]) ? dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0) : dispatch_get_main_queue();
dispatch_async(queueForSetImage, ^{
    [sself sd_setImage:targetImage imageData:targetData basedOnClassOrViaCustomSetImageBlock:setImageBlock];
    dispatch_main_async_safe(^{
        if (!sself) {
            return;
        }
        [sself sd_setNeedsLayout];
        if (completedBlock && finished) {
            completedBlock(image, error, cacheType, url);
        }
    });
});
```